### PR TITLE
Live Preview: "Install&Preview" themes on Atomic sites 

### LINKED ISSUE #80089: Live Preview: Pre-install theme on Atomic sites when selecting the Live Preview option
## Description

As proposed in p9o2xV-3jG-p2#comment-7963, when users select the Live Preview option in Atomic sites, we will install the theme first and then redirect users to the BTP in the Site Editor. This is the same approach used for the "Activate" and "Try & Customize" options.

Note that we need to be careful about upgrade-needed/subscribe-needed themes. 

## Acceptance Criteria

- [x] Implement the logic described above for Atomic sites.

```[tasklist]
### Tasks
- [ ] https://github.com/Automattic/wp-calypso/pull/80353
```

### PR DESCRIPTION
**Pull request by okmttdhr**
Close #80089
Related to https://github.com/Automattic/wp-calypso/pull/80163, https://github.com/WordPress/gutenberg/issues/50749

## Proposed Changes

This PR allows Atomic sites to preview non-installed themes by installing the previewing theme on demand. 

- This PR is only for Atomic sites and NOT for Simple sites. We can address it for Simple sites in the later iteration. (c.f. https://github.com/Automattic/wp-calypso/issues/80411)
- This PR is NOT for externally managed themes (= partner themes), as we should not install such a theme if a user is not subscribed to the theme. We need to think about another solution for externally managed themes.

## Testing Instructions

- Create an Atomic site
- Go to the Live Preview by clicking the "Live Preview" on the Theme Detail page (or clicking the theme option on the Theme Showcase page)
- Verify that you can preview non-installed themes, such as;
	- the Site Editor compatible free non-installed themes
	- the Site Editor compatible premium/woo non-installed themes (e.g. Yuga, Tazza)
	- the Site Editor compatible wporg non-installed themes (e.g. Crossfit Block)
	- the Site Editor compatible Partner INSTALLED themes (e.g. Bonum, Nokul, Macchiato)
		- You should not see the "Live Preview" for non-installed partner themes

https://github.com/Automattic/wp-calypso/assets/5287479/8892bac6-a5b0-4753-8d0c-24fd54e0fed5

https://github.com/Automattic/wp-calypso/assets/5287479/0c37131c-c716-4e21-a2cb-bebce9a0b916

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8698660

Hi @okmttdhr, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Preparing the live preview…`
- `Live preview`

Thank you in advance!

File: client/my-sites/theme/live-preview-button/index.tsx [hunk 1]
```diff
@@ -16,12 +17,18 @@ interface Props {
...
 	);
-	const livePreviewUrl = useSelector( ( state ) => getLivePreviewUrl( state, themeId, siteId ) );
 	if ( ! isLivePreviewSupported ) {
 		return null;
 	}
-	return <Button href={ livePreviewUrl }>{ translate( 'Live Preview' ) }</Button>;
+
+	return (
+		<Button onClick={ () => dispatch( livePreview( themeId, siteId ) ) }>
+			{ translate( 'Live Preview' ) }
```
Comment (fushar) [inline client/my-sites/theme/live-preview-button/index.tsx]: Nit: I think this should be `Live preview` with lowercase `p` to be consistent with the rest of the site 😄

File: client/my-sites/themes/theme-options.js [hunk 2]
```diff
@@ -251,7 +251,7 @@ function getAllThemeOptions( { translate, isFSEActive } ) {
 		label: translate( 'Live Preview', {
```
Comment (fushar) [inline client/my-sites/themes/theme-options.js]: Same here, should be `Live preview`

File: client/my-sites/themes/thanks-modal.jsx [hunk 3]
```diff
@@ -235,7 +236,12 @@ class ThanksModal extends Component {
 	};
 
 	getEditSiteLabel = () => {
-		const { shouldEditHomepageWithGutenberg, hasActivated, isFSEActive } = this.props;
+		const { shouldEditHomepageWithGutenberg, hasActivated, isFSEActive, isLivePreviewStarted } =
+			this.props;
+
+		if ( isLivePreviewStarted ) {
+			return this.props.translate( 'Preparing the Live Preview…' );
```
Comment (fushar) [inline client/my-sites/themes/thanks-modal.jsx]: I can't see this text in the Live Preview flow, how to see it?

**Review by fushar — COMMENTED** [review]
Nice work! I see one error, not sure why, can investigate more tomorrow

### Already installed themes

|Test case|Tested with theme|Result|
|-|-|-|
|Active theme|Artly| ✅ LP menu item not shown|
|Free theme|Didone| ✅ |
|Free "auto-loading-homepage" theme |Heiwa|✅ LP button/menu item not shown|
|Premium theme|Yuga|✅|
|Partner theme|Nokul|✅ |
|Community theme|Stacks| ✅ |

### Non-installed themes

|Test case|Tested with theme|Result|
|-|-|-|
|Free theme|Issue| ✅ |
|Free "auto-loading-homepage" theme |Appleton|✅ LP button/menu item not shown|
|Premium theme|Cultivate|✅|
|Partner theme|Beleco|✅ LP button/menu item not shown|
|Community theme|Crossfit Block| ❌ ERROR, see below|

image

```
Error Details
=============
An error of type E_ERROR was caused in line 113 of the file /srv/htdocs/wp-content/plugins/editing-toolkit-1/starter-page-templates/class-starter-page-templates.php. Error message: Uncaught Error: Failed opening required '/srv/htdocs/wp-content/plugins/editing-toolkit-1/starter-page-templates/class-wp-rest-sideload-image-controller.php' (include_path='/:.') in /srv/htdocs/wp-content/plugins/editing-toolkit-1/starter-page-templates/class-starter-page-templates.php:113
Stack trace:
#0 /wordpress/core/6.3/wp-includes/class-wp-hook.php(310): A8C\FSE\Starter_Page_Templates->register_rest_api(Object(WP_REST_Server))
#1 /wordpress/core/6.3/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters(NULL, Array)
#2 /wordpress/core/6.3/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#3 /wordpress/core/6.3/wp-includes/rest-api.php(587): do_action('rest_api_init', Object(WP_REST_Server))
#4 /wordpress/core/6.3/wp-includes/rest-api.php(545): rest_get_server()
#5 /wordpress/core/6.3/wp-includes/rest-api.php(2911): rest_do_request(Object(WP_REST_Request))
#6 [internal function]: rest_preload_api_request(Array, '/wp/v2/media')
#7 /wordpress/core/6.3/wp-includes/block-editor.php(743): array_reduce(Array, 'rest_preload_ap...', Array)
#8 /wordpress/core/6.3/wp-admin/site-editor.php(116): block_editor_rest_api_preload(Array, Object(WP_Block_Editor_Context))
#9 {main}
  thrown
```

**Comment by fushar** [comment]
This is weird, when I test Crossfit Blcok again, using the same site and another site, it worked... not sure what's happening.

|Test case | Tested with theme | Result|
|-|-|-|
|Community theme | Crossfit Block | ✅ |

**Comment by miksansegundo** [comment]
Nice! We are getting closer 🙌 

The first time worked but the second time failed with the error `theme_already_installed` in the request `public-api.wordpress.com/rest/v1.1/sites/{ SITE }/themes/montagna/install` and no redirection when I click `Live Preview`.   

Screenshot 2566-08-11 at 17 57 09

**Comment by okmttdhr** [comment]
> This is weird, when I test Crossfit Blcok again, using the same site and another site, it worked... not sure what's happening.
> 
> Test case	Tested with theme	Result
> Community theme	Crossfit Block	✅

This is a bit scary 👀  Let's keep an eye on and investigate this if that happens again.

**Comment by okmttdhr** [comment]
> The first time worked but the second time failed with the error `theme_already_installed` in the request `public-api.wordpress.com/rest/v1.1/sites/{ SITE }/themes/montagna/install` and no redirection when I click `Live Preview`.

Thanks for your testing, @miksansegundo! I addressed this, I'd appreciate it if you try it again when it's convenient for you!

**Review by fushar — APPROVED** [review]
I've re-tested this on Free, Premium, and Business (non-Atomic & Atomic) sites again, and I think everything works great. Nice tests on `getIsLivePreviewSupported()` btw.

Thanks for working on this!

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8698660).

**Pull request merged**